docs: rewrite and restructure meet-motoko.md [DRAFT] (#4885)
dfinity/motoko: doc/md/meet-motoko.md

sidebar_position: 1
---

> # Meet Motoko: The language of the Internet Computer

Reviewer: What about just: 

Meet Motoko: a language designed for the Internet Computer.

Motoko is the first programming language dedicated to the Internet Computer....

@jessiemongeon1

> # Meet Motoko: The language of the Internet Computer

Reviewer 2: I'd still like to keep it general. Maybe just "Getting started with Motoko" or "Motoko introduction and setup"

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

import Tabs from "@theme/Tabs";

import TabItem from "@theme/TabItem";

> <MarkdownChipRow labels={["Beginner", "Getting started"]} />

Reviewer 2: Note for self: change pages to MarkdownX to support tabs & other custom TS

<img src="https://github.com/user-attachments/assets/844ca364-4d71-42b3-aaec-4a6c3509ee2e" alt="Logo" width="150" height="150" />

## Introduction

> Motoko is the native programming language of the Internet Computer. It’s a modern, **actor-based** language designed for safety, efficiency, and scalability. With native support for asynchronous messaging, WebAssembly (Wasm) interoperability, robust type safety, and data persistence, Motoko makes it easy to build secure and performant decentralized web applications.

Reviewer: ```suggestion
Motoko is the native programming language of the Internet Computer. It’s a modern, **actor-based** language designed for safety, efficiency, and scalability. With native support for asynchronous messaging, Candid interoperability, robust type safety, and data persistence, Motoko makes it easy to build secure and performant decentralized web applications.
```
Might be better. Wasm interop tends to sugges the wasm component model which neither Motoko nor the IC supports yet.

> Motoko is the native programming language of the Internet Computer. It’s a modern, **actor-based** language designed for safety, efficiency, and scalability. With native support for asynchronous messaging, WebAssembly (Wasm) interoperability, robust type safety, and data persistence, Motoko makes it easy to build secure and performant decentralized web applications.

Reviewer 2: ```suggestion
Motoko is a modern, actor-based language designed for safety, efficiency, and scalability. Motoko natively supports the unique features of the Internet Computer, including asynchronous messaging, robust type safety, and automatic memory management.
```

Reviewer: @jessiemongeon1 neither type safety or automatic memory management are unique features of the IC (not asynchronous messaging for that matter). The original wording was actually more precise.

Reviewer 2: @crusso Revised, let me know if that is better

## Install Motoko

Follow these steps to set up your Motoko development environment:

> 1. **Install the IC SDK (DFX)**

Reviewer 2: ```suggestion
1. **Install the IC SDK**
```

Reviewer 2: To the reader, there is no explanation as to why the IC SDK is required for Motoko development. This needs some brief explanation such as:

> To develop Motoko applications, you will need a developer environment that contains the Motoko compiler and base library. It is recommended to use the [IC SDK](https://github.com/dfinity/sdk#readme), a CLI tool that comes packaged with the Motoko base library and compiler.

  ```bash
  sh -ci "$(curl -fsSL https://internetcomputer.org/install.sh)"
  ```

  Verify the installation with:

  ```bash
  dfx --version
  ```

2. **Install Visual Studio Code**
   - Download VSCode from [https://code.visualstudio.com](https://code.visualstudio.com)

>    - Follow the installation instructions for your operating system

Reviewer 2: This isn't necessary, and if developers use another IDE, they will be hesitant to download a new one that they are not familiar with.

> 2. **Install Visual Studio Code**
>    - Download VSCode from [https://code.visualstudio.com](https://code.visualstudio.com)
>    - Follow the installation instructions for your operating system

Reviewer 2: ```suggestion

```

> 3. **Install Motoko Extension**
>    - Open VSCode
>    - Go to Extensions (Ctrl+Shift+X / Cmd+Shift+X)
>    - Search for "Motoko"
>    - Install the official Motoko extension

Reviewer 2: ```suggestion

```

Reviewer 2: Also not necessary, but we should instead mention that the VS Code extension exists and is recommended for VS Code users.

## Explore Motoko use cases

<Tabs>

> <TabItem value="usecase1" label="Backend" default>

Reviewer 2: "Backend" is not really a "use case" but rather a way to describe where the application's logic is stored.

Reviewer 2: I would rather say "Web applications" as the use case, then provide examples of web application backends.

Reviewer: https://github.com/ninegua/reversi is quite a nice worked example (thought a bit old now).  If too old, perhaps there are others.

Author: > https://github.com/ninegua/reversi is quite a nice worked example (though a bit old now). If too old, perhaps there are others, like Luc's auction

The age i would say doesn't matter, as long as it is syntactically and semantically correct and it  is runnable i would include it. Would definitely add other examples as per your suggestion.

Motoko makes it easy to build scalable backend services for web applications.

1. [Deploy your first Motoko canister.](https://internetcomputer.org/docs/current/tutorials/developer-journey/level-1/1.1-live-demo)

2. [Expand your application by calling external APIs from a Motoko canister.](https://internetcomputer.org/docs/current/developer-docs/backend/https-outcalls)

> 3. [Connect your Motoko backend with a frontend canister.](https://internetcomputer.org/docs/current/references/samples/svelte/svelte-motoko-starter/)  

Reviewer 2: These three examples don't feel like they really correlate with each other or provide comparable examples, they are a variety of different content forms (tutorial, example, etc.) and different levels of complexity. 

I would suggest instead linking to three examples from the dfinity/examples repo (so the format is consistent) and picking three that demonstrate an "exciting" use case. These three examples aren't really good introductions to Motoko; we want the reader to be excited and inspired by viewing these.

</TabItem>

<TabItem value="usecase2" label="DeFi" default>

Build decentralized finance (DeFi) applications using Motoko.

1. [Set up an ICRC-1 canister.](https://internetcomputer.org/docs/current/tutorials/hackathon-prep-course/integrating-with-tokens/#deploying-an-icrc-1-ledger-locally)

> 2. [Create a decentralized exchange (DEX)](https://internetcomputer.org/docs/current/tutorials/developer-journey/level-5/5.3-token-swap-tutorial/)  

Reviewer 2: Same comment from above; change these examples to be ones from the dfinity/examples repo.

</TabItem>

<TabItem value="usecase3" label="Chain Fusion" default>

Develop cross-chain applications and integrate with other blockchains.  

1. [Explore Bitcoin integration (ckBTC).](https://internetcomputer.org/docs/current/developer-docs/multi-chain/bitcoin/overview)
  
2. [Explore Ethereum integration (ckETH).](https://internetcomputer.org/ethereum-integration/)
  

> 3. [Explore Chain-Key ERC-20 (ckERC-20) integrations.](https://internetcomputer.org/docs/current/developer-docs/multi-chain/chain-key-tokens/ckerc20/overview/)

Reviewer 2: Same comment as above; change these examples to be ones from the dfinity/examples repo.

Also, the Bitcoin/Ethereum integrations are not just ckBTC/ckETH, those are just a small portion of the integrations. I would avoid mentioning ckTokens here, as they used for DeFi applications rather than cross-chain applications.

Author: I have made changes addressing the feedback including a title revert, improved IC SDK explanation, updated use case links, refined IDE recommendations, removed digital assets section. let me know what you think

</TabItem>

> <TabItem value="usecase4" label="Digital assets" default>
> 
> Manage digital assets on the Internet Computer.
> 
> 1. [Learn how to deploy NFTs on ICP](https://internetcomputer.org/docs/current/tutorials/developer-journey/level-5/5.4-NFT-tutorial)
> 
> </TabItem>

Reviewer 2: ```suggestion

```

Reviewer 2: This is redundant to DeFi and can be removed.

</Tabs>

> ## Quick References  

Reviewer 2: ```suggestion
## Quick references
```

For developers looking for quick access to key resources, here are direct links to the base library and useful examples:

> - **[Motoko base library](https://internetcomputer.org/docs/current/motoko/main/base/)** – Browse core modules and functions.  

Reviewer 2: ```suggestion
- **[Motoko base library](https://internetcomputer.org/docs/current/motoko/main/base/)**

> - **[Examples:](https://internetcomputer.org/docs/current/motoko/main/getting-started/examples)**

Reviewer 2: ```suggestion
- **[Motoko example projects](https://github.com/dfinity/examples/tree/master/motoko)**
```

### Share your feedback

> Was this page helpful? [Yes] [No]

Reviewer 2: We currently do not have any way to collect feedback in this format. What was the intention behind adding this?

Motoko is a modern, **actor-based** language designed for safety, efficiency, and scalability. Motoko natively supports asynchronous messaging, Candid interoperability, robust type safety, automatic memory management, and the unique features of the Internet Computer.

## Install Motoko

Follow these steps to set up your Motoko development environment:

### 1. **Install the IC SDK**  

To develop Motoko applications, you need a development environment that includes the Motoko compiler and base library. The Internet Computer Software Development Kit (IC SDK) provides these essential tools, along with utilities for managing and deploying canisters.  

```bash
sh -ci "$(curl -fsSL https://internetcomputer.org/install.sh)"
```

Verify the installation with:

```bash
dfx --version
```

### 2. **Choose an IDE**  

While you can use any IDE for Motoko development, **Visual Studio Code (VS Code)** is recommended due to its convenient extension support. If you prefer to use VS Code, you can download it from:  
[https://code.visualstudio.com](https://code.visualstudio.com)  

> ### 3. **Install the Motoko Extension(Optional, Recommended for VS Code Users)**

Reviewer 2: ```suggestion
### 3. **Install the Motoko VS Code extension (optional, recommended for VS Code Users)**
```

If you are using VS Code, you can enhance your development experience by installing the official Motoko extension:  

> - Open VS Code  
> - Go to Extensions (`Ctrl+Shift+X` / `Cmd+Shift+X`)  
> - Search for **"Motoko"**  
> - Install the official Motoko extension  

Reviewer 2: ```suggestion
1. Open VS Code.
2. Go to "Extensions" (`Ctrl+Shift+X` / `Cmd+Shift+X`).
3. Search for **"Motoko"**
4. Install the official Motoko extension.
```

## Explore Motoko use cases

<Tabs>

<TabItem value="usecase1" label="Web applications" default>

Motoko makes it easy to build scalable backend services for web applications.

> 1. [Deploy your first Motoko canister.](https://github.com/dfinity/examples/tree/master/motoko/counter)

Reviewer 2: Note for self: use ICP Ninja URL in future when available

2. [Expand your application by calling external APIs from a Motoko canister.](https://github.com/dfinity/examples/tree/master/motoko/send_http_get)

> 3. [Explore a Reversi game built in motoko.](https://github.com/ninegua/reversi)  

Reviewer 2: ```suggestion
3. [Explore a Reversi game built in Motoko.](https://github.com/ninegua/reversi)  
```

</TabItem>

<TabItem value="usecase2" label="DeFi" default>

Build decentralized finance (DeFi) applications using Motoko.

1. [Set up an ICRC-1 canister.](https://github.com/sonicdex/icrc-1-public/)

> 2. [Create a decentralized exchange (DEX)](https://github.com/dfinity/examples/tree/master/motoko/icrc2-swap)  

Reviewer 2: ```suggestion
2. [Create a decentralized exchange (DEX).](https://github.com/dfinity/examples/tree/master/motoko/icrc2-swap)  
```

> 3. [Learn how to deploy NFTs on ICP](https://github.com/noku-team/icrc7_motoko)

Reviewer 2: ```suggestion
3. [Learn how to deploy NFTs on ICP.](https://github.com/noku-team/icrc7_motoko)
```

</TabItem>

<TabItem value="usecase3" label="Chain Fusion" default>

Develop cross-chain applications and integrate with other blockchains.  

> 1. [Explore Ethereum integration](https://github.com/dfinity/icp-eth-starter)

Reviewer 2: ```suggestion
1. [Explore the Ethereum integration.](https://github.com/dfinity/icp-eth-starter)
```
